### PR TITLE
Add rgcn-authentication detection

### PR DESCRIPTION
**Pull request by tzemicheal**
This closes
https://github.com/nv-morpheus/Morpheus/issues/218

### CODE REVIEW

**Review by BartleyR — COMMENTED** [review]
The code (specifically the `.py` files) needs more inline documentation.

**Comment by shawn-davis** [comment]
lgtm

**Review by raykallen — CHANGES_REQUESTED** [review]
for the notebook: intro needs to include why rgcn over traditional graph methods and rules-based methods. What is the motivation?
Include rational for using the status flag for training
List the features used for OHE in the authentication node and lack of OHE features in other nodes
Graph the results for the synthetic dataset. Discuss what makes them different from the internal dataset. Suggest what attributes make a good dataset for this technique.

**Comment by raykallen** [comment]
Please add blurb with link on the main repo readme here: https://github.com/nv-morpheus/morpheus-experimental#current-cybersecurity-workflow-prototypes

File: malicious-authentication-detection/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#### GPU model
+V100
+
+#### Model accuracy
+AUC = 0.78 (RGCN binary classifier)
+
+
+### How To Use This Model
+This model is an example of a authentication detection pipeline using a graph neural network. This can be further retrained or fine-tuned to be used for similar types of  networks with similar graph structures.
+
```
[the remark this answers is not in the document]
Comment (BartleyR) [reply malicious-authentication-detection/README.md:42]: Yes. Add any pre-reqs for a dataset that would make it amenable to this approach.

File: malicious-authentication-detection/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+## Use Case
+Detection of Malicious Accounts on Azure-AD signon using Relational Graph Neural Network (RGCN)
+
+### Version
+1.0
+
+### Model Overview
+
+This model shows an application of a graph neural network for malicious authentication detection in Azure-AD signon heterogeneous graph. An Azure-AD signon dataset  includes four types of nodes, authentication, user, device and service application nodes are used for modeling. A Relational graph neural network (RGCN)  is used to identify malicious authentications.
```
Comment (BartleyR) [inline malicious-authentication-detection/README.md:11]: Is it possible to link to more information about Azure-AD online? Are there resource pages from MSFT?

File: malicious-authentication-detection/README.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+A training data consists of 45K authentication event, with semisupervised settings. The dataset is extracted from Azure-AD sign on events. The RGCN is trained to output embedded representation of authentication out of the graph and binary classification of individual authentication. 
+#### Training epochs
+30
+
+#### Training batch size
+1000
+
+#### GPU model
+V100
```
Comment (BartleyR) [inline malicious-authentication-detection/README.md:34]: Which version?

**Review by BartleyR — CHANGES_REQUESTED** [review]
Comments on the notebook:

- The notebook is missing a motivation - i.e., why did we even do this in the first place? Although this may be better placed in the `README.md`. Unless you have a preference @raykallen , I'd like this placed in the `README.md` and can just be summarized/replicated in the notebook.
- Why not make the references hot-linked? Would be much easier to follow them.
- Write a few sentences explaining the image (results) shown in the Evaluation section

File: malicious-authentication-detection/README.md (hunk 2 above)
Comment (BartleyR) [reply malicious-authentication-detection/README.md:11]: I see the link in the notebook. We should make sure we're presenting all of the salient information on GitHub, since that will be the first landing site.

File: malicious-authentication-detection/README.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+### Version
+1.0
+
+### Model Overview
+
+This model shows an application of a graph neural network for malicious authentication detection in Azure-AD signon heterogeneous graph. An Azure-AD signon dataset  includes four types of nodes, authentication, user, device and service application nodes are used for modeling. A Relational graph neural network (RGCN)  is used to identify malicious authentications.
+
+### Model Architecture
+It uses a  heterogeneous graph representation as input for RGCN. Since the input graph is heterogenous, an embedding for target node "authentication" is used for training the classifier.
+
```
[the remark this answers is not in the document]
Comment (tzemicheal) [reply malicious-authentication-detection/README.md:15]: I will update the readme.

File: malicious-authentication-detection/README.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+### How To Use This Model
+This model is an example of a authentication detection pipeline using a graph neural network. This can be further retrained or fine-tuned to be used for similar types of  networks with similar graph structures.
+
+### Input
+Authentication data with nodes including user, authentication, device and service.
+
+### Output
+An anomalous score of authentication indicates a probability score of being malicious
+
```
[the remark this answers is not in the document]
Comment (tzemicheal) [reply malicious-authentication-detection/README.md:48]: A threshold of 0.499 is used to output failure/success. It has the option to output as `test_log.argmax(dim=1)`

File: malicious-authentication-detection/README.md (hunk 3 above)
Comment (tzemicheal) [reply malicious-authentication-detection/README.md]: Fixed this in the new revision.

File: malicious-authentication-detection/README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
+## Detection of Malicious authentication  using Relational Graph Neural Network (RGCN)
+
+## Use Case
+Detection of Malicious Accounts on Azure-AD signon using Relational Graph Neural Network (RGCN)
```
Comment (hsin-c) [inline malicious-authentication-detection/README.md:4]: Are we detecting malicious account or authentication? (or both? the title suggests authentication but this line suggests otherwise)
Also, is this model detection "malicious" behavior or "anomalous" behavior? (or error-prone behavior?) We seem to be using those interchangeably in this readme, but I feel like detecting "malicious" is a much heavier lift as security expertise would often be required (hence hard to achieve by unsupervised or semisupervised methods).

Comment (tzemicheal) [reply malicious-authentication-detection/README.md:4]: > Are we detecting malicious account or authentication? (or both? the title suggests authentication but this line suggests otherwise) Also, is this model detection "malicious" behavior or "anomalous" behavior? (or error-prone behavior?) We seem to be using those interchangeably in this readme, but I feel like detecting "malicious" is a much heavier lift as security expertise would often be required (hence hard to achieve by unsupervised or semisupervised methods).

The primary task is the detection of malicious authentication, all the metric is done on authentication. For user detection, the authentications are aggregated across authentication made by a user(e.g. in the experiment I took the median scores of all authentication made for each user per timestamp), based on this user detection is a summarized part of the authentication detection. 

For the second point, the goal of this use case is to detect malicious users. I agree we have been using the word interchangeable. In my opinion, malicious behavior is a subset of anomalous behavior. Technically, we trained the model to identify "failure" cases, which could be either abnormal or malicious. But, the evaluation is done on test data that is simulated to act as malicious behavior (synthetic dataset) or internal result reported.

File: malicious-authentication-detection/training-tuning/model.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import dgl.function as fn
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class HeteroRGCNLayer(nn.Module):
```
[the remark this answers is not in the document]
Comment (tzemicheal) [reply malicious-authentication-detection/training-tuning/model.py:22]: Fixed this script.

File: malicious-authentication-detection/README.md (hunk 6 above)
Comment (BartleyR) [reply malicious-authentication-detection/README.md:4]: This is more anomalous (like Hsin mentioned) than malicious, right? Something can be anomalous and not be malicious. But I don't think we're specifically classifying intent here.

Comment (tzemicheal) [reply malicious-authentication-detection/README.md]: Updated the README/NB. Thanks.

**Review by raykallen — APPROVED** [review]
lgtm

**Review by BartleyR — APPROVED** [review]
LGTM

**Comment by raykallen** [comment]
@gpucibot merge
